File: packages/tactical-engine-py/tactical_engine/targeting/resolve_line.py
from shared_contracts import Coordinate, Obstacle
from ..validation.obstacle_rules import blocks_targeting


def resolve_line(
    origin: Coordinate,
    anchor: Coordinate,
    range_: int,
    obstacles: list[Obstacle],
) -> list[Coordinate]:
    result: list[Coordinate] = []
    delta_x = _sign(anchor.x - origin.x)
    delta_y = _sign(anchor.y - origin.y)

    for step in range(1, range_ + 1):
        coordinate = Coordinate(x=origin.x + delta_x * step, y=origin.y + delta_y * step)
        if blocks_targeting(obstacles, coordinate):
            break
        result.append(coordinate)

    return result


def _sign(v: int) -> int:
    if v > 0:
        return 1
    if v < 0:
        return -1
    return 0

File: packages/tactical-engine/src/targeting/resolve-line.js
import { blocksEffect } from "../validation/obstacle-rules";
import { findEdgeBetween } from "../grid/grid-state";
export function resolveLine(origin, anchor, range, obstacles, edgeObstacles = []) {
    const result = [];
    const deltaX = Math.sign(anchor.x - origin.x);
    const deltaY = Math.sign(anchor.y - origin.y);
    let prev = origin;
    for (let step = 1; step <= range; step += 1) {
        const coordinate = { x: origin.x + deltaX * step, y: origin.y + deltaY * step };
        if (blocksEffect(obstacles, coordinate))
            break;
        if (edgeObstacles.length > 0) {
            const crossedEdge = findEdgeBetween(edgeObstacles, prev, coordinate);
            if (crossedEdge?.blocksEffect)
                break;
        }
        result.push(coordinate);
        prev = coordinate;
    }
    return result;
}
